universal-login-relayer/src/integration/ethereum/MessageExecutor.ts

```diff
@@ -16,15 +18,36 @@ export class MessageExecutor implements IExecutor<SignedMessage> {
     private messageRepository: IMessageRepository,
     private minedTransactionHandler: IMinedTransactionHandler,
     private walletContractService: WalletContractService,
+    private gasTokenValidator: GasTokenValidator,
   ) {}
 
   canExecute(item: QueueItem): boolean {
     return item.type === 'Message';
   }
 
+  async validateMessageItem(messageItem: MessageItem) {
+    const signedMessage = messageItem.message;
+    const {gasPrice, gasToken} = signedMessage;
+    let {tokenPriceInEth} = messageItem;
+    tokenPriceInEth = tokenPriceInEth || '1';
+    this.gasTokenValidator.validate({
+      gasPrice: gasPrice.toString(),
+      gasToken,
+      tokenPriceInETH: tokenPriceInEth,
+    }, 0.3);
+  }
+
+  calculateSignedMessageGasPrice(signedMessage: SignedMessage, messageItem: MessageItem) {
+    const gasPrice = utils.bigNumberify(signedMessage.gasPrice);
+    return safeDivide(gasPrice, messageItem.tokenPriceInEth || '1');
+  }
+
   async handleExecute(messageHash: string) {
     try {
-      const signedMessage = await this.messageRepository.getMessage(messageHash);
+      const messageItem = await this.messageRepository.get(messageHash);
+      this.validateMessageItem(messageItem);
+      const signedMessage = messageItem.message;
+      signedMessage.gasPrice = this.calculateSignedMessageGasPrice(signedMessage, messageItem);
```

Comment:
In my opinion calculating should be done walletContractService.messageToTransaction =>  transactionGasPriceComputator.getGasPriceInEth

Reply:
👍